### PR TITLE
ffmpeg hls guide

### PR DESCRIPTION
**Pull request by mrbar42**

### CODE REVIEW

File: docs/guides/production-ready-hls-vod.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+## Installing FFMPEG
+
+`ffmpeg` is a cross platform program that can run on Windows and OS X as well as Linux.
+
+### Windows
+- Download latest version from [here](http://ffmpeg.zeranoe.com/builds/)
+- Unzip the archive to a folder
+- Open a command prompt in the unzipped folder
+- run `./ffmpeg` - you should see ffmpeg version and build information
```
Comment (benjamingr) [inline docs/guides/production-ready-hls-vod.md]: *R*un

File: docs/guides/production-ready-hls-vod.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+### OS X
+- Install [homebrew](https://brew.sh/)
+- Run `brew install ffmpeg` (extra options can be seen by running `brew options ffmpeg`)
+- run `./ffmpeg` - you should see `ffmpeg` version and build information
```
Comment (benjamingr) [inline docs/guides/production-ready-hls-vod.md]: *R*un

File: docs/guides/production-ready-hls-vod.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+## Multi Bitrate Conversion
+
+### First rendition
+
+Since end users have different screen sizes and different network performance, we want to create multiple renditions of
+the video with different resolutions and bitrates that can be switched seamlessly, this concept is called MBR (Multi Bit Rate).  
+`ffmpeg` expects parameters for every rendition you want to create.
+
+Lets build a command for one rendition
```
Comment (benjamingr) [inline docs/guides/production-ready-hls-vod.md]: End with `:`?

File: docs/guides/production-ready-hls-vod.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+- `-crf 20` - Constant Rate Factor, high level factor for overall quality
+- `-g 48 -keyint_min 48` - **IMPORTANT** create key frame (I-frame) every 48 frames (~2 seconds) - will later affect correct slicing of segments and alignment of renditions
+- `-sc_threshold 0` - don't create  key frames on scene change - only according to `-g`
+- `-b:v 2400k -maxrate 2500k -bufsize 4800k` - limit video bitrate, these are rendition specific and depends on your content type - [read more](#how-to-choose-the-right-bitrate)
+- `-hls_time 4` - segment target duration in seconds - the actual length is constrained by key frames
+- `-hls_playlist_type vod` - adds the `#EXT-X-PLAYLIST-TYPE:VOD` tag and keeps all segments in the playlist
+- `-hls_segment_filename beach/720p_%03d.ts` - explicitly define segments files names
+- `beach/720p.m3u8` - path of the playlist file - also tells ffmpeg to output HLS (`.m3u8`)
+
+this will generate a VOD HLS playlist and segments in `beach` folder
```
Comment (benjamingr) [inline docs/guides/production-ready-hls-vod.md]: *T*his and `.`

File: docs/guides/production-ready-hls-vod.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+### Multiple renditions
+
+each rendition can be generated separately or they can all be generated in parallel with one command.
```
Comment (benjamingr) [inline docs/guides/production-ready-hls-vod.md]: `E`ach

File: docs/guides/production-ready-hls-vod.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+- `-hls_playlist_type vod` - adds the `#EXT-X-PLAYLIST-TYPE:VOD` tag and keeps all segments in the playlist
+- `-hls_segment_filename beach/720p_%03d.ts` - explicitly define segments files names
+- `beach/720p.m3u8` - path of the playlist file - also tells ffmpeg to output HLS (`.m3u8`)
+
+this will generate a VOD HLS playlist and segments in `beach` folder
+
+### Multiple renditions
+
+each rendition can be generated separately or they can all be generated in parallel with one command.  
+its very important that besides the resolution and bitrate parameters the command will be identical so that the renditions will be properly aligned,
```
Comment (benjamingr) [inline docs/guides/production-ready-hls-vod.md]: It's, command*s*

File: docs/guides/production-ready-hls-vod.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+meaning key frames will be set in the exact same positions to allow smooth switching between them on the fly.
+
+I will create 4 renditions with common resolutions:
```
Comment (benjamingr) [inline docs/guides/production-ready-hls-vod.md]: We, or Let's like above

File: docs/guides/production-ready-hls-vod.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  -vf scale=w=842:h=480:force_original_aspect_ratio=decrease -c:a aac -ar 48000 -c:v h264 -profile:v main -crf 20 -sc_threshold 0 -g 48 -keyint_min 48 -hls_time 4 -hls_playlist_type vod -b:v 1400k -maxrate 1498k -bufsize 2100k -b:a 128k -hls_segment_filename beach/480p_%03d.ts beach/480p.m3u8 \
+  -vf scale=w=1280:h=720:force_original_aspect_ratio=decrease -c:a aac -ar 48000 -c:v h264 -profile:v main -crf 20 -sc_threshold 0 -g 48 -keyint_min 48 -hls_time 4 -hls_playlist_type vod -b:v 2800k -maxrate 2996k -bufsize 4200k -b:a 128k -hls_segment_filename beach/720p_%03d.ts beach/720p.m3u8 \
+  -vf scale=w=1920:h=1080:force_original_aspect_ratio=decrease -c:a aac -ar 48000 -c:v h264 -profile:v main -crf 20 -sc_threshold 0 -g 48 -keyint_min 48 -hls_time 4 -hls_playlist_type vod -b:v 5000k -maxrate 5350k -bufsize 7500k -b:a 192k -hls_segment_filename beach/1080p_%03d.ts beach/1080p.m3u8
+</pre>
+
+### Master playlist
+
+The HLS player needs to know that there are multiple renditions of our video, so we create an HLS master playlist to point them,
+and save it along side the other playlists and segments.
```
Comment (benjamingr) [inline docs/guides/production-ready-hls-vod.md]: excess comma

File: docs/guides/production-ready-hls-vod.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+480p.m3u8
+#EXT-X-STREAM-INF:BANDWIDTH=2800000,RESOLUTION=1280x720
+720p.m3u8
+#EXT-X-STREAM-INF:BANDWIDTH=5000000,RESOLUTION=1920x1080
+1080p.m3u8
+```
+
+### Example Conversion Script
+
+here is an example conversion script [create-hls-vod.sh](https://gist.github.com/mrbar42/ae111731906f958b396f30906004b3fa)
```
Comment (benjamingr) [inline docs/guides/production-ready-hls-vod.md]: Capital here and the one below

File: docs/guides/production-ready-hls-vod.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+      |- 1080p.m3u8
+      |- 1080p_001.ts
+      |- 1080p_002.ts  
+```
+
+# FAQ
+
+## How to choose the right bitrate
+
+bitrate is dependant on the resolution and the content type. a video with a lot of movement and rapid changes in picture
```
Comment (benjamingr) [inline docs/guides/production-ready-hls-vod.md]: Capital, ... of the video

Comment (benjamingr) [reply docs/guides/production-ready-hls-vod.md]: The paragraph needs rephrasing

File: docs/guides/production-ready-hls-vod.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
+# Creating A Production Ready Multi Bitrate HLS VOD stream
+
+[HLS](https://en.wikipedia.org/wiki/HTTP_Live_Streaming) is one of the most prominent video streaming formats on desktop and mobile browsers. `ffmpeg` is a powerful tool that supports conversion of various video formats from one to another, including HLS both as input and output.
```
Comment (shacharz) [inline docs/guides/production-ready-hls-vod.md]: Start with talking abut multibitrate and what is it good for, before diving into ffmpeg.
You can take the paragraphs from 'first rendition' and elaborate upon it.

**Pull request merged**